bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/RowResultAdapter.java

```diff
@@ -0,0 +1,224 @@
+      // because value(), size(), isEmpty() rawCells() etc. are directly using Result's cells field.
+      this.copyFrom(Result.create(cells));
+    }
+
+    ByteString getKey() {
+      return rowKey;
+    }
+
+    boolean isMarkerRow() {
+      return isMarkerRow;
+    }
+  }
+
+  public static class RowResultBuilder implements RowBuilder<Result> {
+    private ByteString currentKey;
+    private String family;
+    private byte[] qualifier;
+    private List<String> labels;
+    private long timestamp;
+    private byte[] value;
+
+    private Map<String, List<RowCell>> cells = new TreeMap<>();
+    private List<RowCell> currentFamilyCells = null;
+    private String previousFamily;
+    private int nextValueIndex;
+
+    @Override
+    public void startRow(ByteString rowKey) {
+      this.currentKey = rowKey;
+    }
+
+    @Override
+    public void startCell(
+        String family, ByteString qualifier, long timestamp, List<String> labels, long size) {
+      this.family = family;
+      this.qualifier = qualifier.toByteArray();
+      this.timestamp = timestamp;
+      this.labels = labels;
+      if (size > 0) {
+        this.value = new byte[(int) size];
+        this.nextValueIndex = 0;
+      } else {
+        this.value = EMPTY_VALUE;
+        this.nextValueIndex = -1;
+      }
+    }
+
+    @Override
+    public void cellValue(ByteString newValue) {
+      // TODO: Verify if can value ever be null?
+      if (newValue == null || newValue.isEmpty()) {
+        return;
+      }
+
+      if (nextValueIndex == -1) {
+        this.value = newValue.toByteArray();
+        nextValueIndex = newValue.size();
+      } else {
+        int newValueSize = newValue.size();
+        Preconditions.checkState(
+            Integer.MAX_VALUE - nextValueIndex > newValueSize,
+            "value would be too large to contain");
+
+        this.value = Arrays.copyOf(this.value, nextValueIndex + newValueSize);
+        System.arraycopy(newValue.toByteArray(), 0, this.value, nextValueIndex, newValueSize);
+        nextValueIndex += newValueSize;
+      }
+    }
+
+    /**
+     * Adds a Cell to {@link RowCell}'s map which is ordered by family. cells received from {@link
+     * RowBuilder} has ordering as:
+     *
+     * <ul>
+     *   <li>family names clustered, but not sorted
+     *   <li>qualifiers in each family cluster is sorted lexicographically
+     *   <li>then descending by timestamp
+     * </ul>
+     *
+     * The end result will be that {@link RowCell} are ordered as:
+     *
+     * <ul>
+     *   <li>lexicographical by family
+     *   <li>then lexicographical by qualifier
+     *   <li>then descending by timestamp
+     * </ul>
+     *
+     * A flattened version of the {@link RowCell} map will be sorted correctly.
+     */
+    @Override
+    public void finishCell() {
+      Preconditions.checkState(
+          nextValueIndex == -1 || nextValueIndex == value.length, "Inconsistent value found");
```

Comment:
nextValueIndex should not be -1. Also the error message is not helpful

reconditions.checkState(nextValueIndex == value.length, "Cell value too short");